tests/contrib/test_integration_registry.py

```diff
@@ -0,0 +1,258 @@
+    if jsonschema is None:
+        pytest.skip("jsonschema library not installed, skipping schema validation.")
+
+    try:
+        jsonschema.validate(instance=registry_content, schema=registry_schema)
+    except jsonschema.ValidationError as e:
+        pytest.fail(f"Schema validation failed for {REGISTRY_YAML_PATH}:\n{e}", pytrace=False)
+    except Exception as e:
+        pytest.fail(f"An unexpected error occurred during schema validation: {e}")
+
+
+def test_all_internal_dirs_accounted_for(registry_data: list[dict], all_integration_names: set[str]):
+    """
+    Verify that every directory within ddtrace/contrib/internal is listed
+    as an 'integration_name' in registry.yaml.
+    """
+    if not CONTRIB_INTERNAL_DIR.is_dir():
+        pytest.fail(f"contrib/internal directory not found: {CONTRIB_INTERNAL_DIR}")
+
+    accounted_for_dirs = all_integration_names
+    found_dirs_on_disk = set()
+    unaccounted_dirs = []
+
+    for item in CONTRIB_INTERNAL_DIR.iterdir():
+        if item.is_dir() and item.name != "__pycache__":
+            dir_name = item.name
+            found_dirs_on_disk.add(dir_name)
+            if dir_name not in accounted_for_dirs:
+                unaccounted_dirs.append(dir_name)
+
+    # Check for entries in YAML that don't have a corresponding directory
+    missing_dirs_in_yaml = accounted_for_dirs - found_dirs_on_disk
+
+    error_messages = []
+    if unaccounted_dirs:
+        error_messages.append(
+            f"\n\nUnaccounted Directories Found:\n"
+            f"    The following directories exist in '{CONTRIB_INTERNAL_DIR.relative_to(PROJECT_ROOT)}' \n"
+            f"    but do NOT have a corresponding 'integration_name' entry in '{REGISTRY_YAML_PATH.name}'.\n"
+            f"    Please add entries for them:\n"
+            f"      - " + "\n      - ".join(sorted(unaccounted_dirs)) + "\n"
+        )
+    if missing_dirs_in_yaml:
+        error_messages.append(
+            f"\n\nMissing Directories for YAML Entries:\n"
+            f"    The following integration names are listed in '{REGISTRY_YAML_PATH.name}'\n"
+            f"    but do not have a corresponding directory in '{CONTRIB_INTERNAL_DIR.relative_to(PROJECT_ROOT)}':\n\n"
+            f"      - " + "\n      - ".join(sorted(list(missing_dirs_in_yaml))) + "\n"
+        )
+
+    assert not error_messages, "\n".join(error_messages)
+
+
+def test_external_package_requirements(registry_data: list[dict]):
+    """
+    Verify that external packages have dependency names and valid version fields.
+    """
+    errors = []
+    version_fields_to_check = ["tested_version_min", "tested_version_max"]
+
+    for entry in registry_data:
+        name = entry.get("integration_name", "UNKNOWN_ENTRY")
+        is_external = entry.get("is_external_package")
+        is_tested = entry.get("is_tested")
+
+        if is_external is True and is_tested is not False:
+            # 1. Check if dependency_name exists
```

Comment:
Putting the code's logical sections into their own functions is a way of organizing that makes the code significantly less nested and more readable than using commented blocks. I recommend splitting this large function into several smaller ones.

Reply:
Will do, agreed that this code (like most of the code for this PR currently), is quite messy. Working on refactoring and cleaning it up now.